flink-streaming-connectors/flink-connector-kafka-base/src/main/java/org/apache/flink/streaming/connectors/kafka/FlinkKafkaProducerBase.java

```diff
@@ -113,6 +123,14 @@
 	/** Errors encountered in the async producer are stored here */
 	protected transient volatile Exception asyncException;
 
+	/**
+	 * Number of unacknowledged records.
+	 * There is no need to introduce additional locks because invoke() and snapshotState() are
+	 * never called concurrently. So blocking the snapshotting will lock the invoke() method until all
+	 * pending records have been confirmed.
+	 */
```

Comment:
I think the fact that `invoke` and `snapshotState` are mutually exclusive is not important for the semantics of the `pendingRecords` variable. The reason is that it will only be incremented in `invoke` and decremented in the `callbacks` of the Kafka producer.